### PR TITLE
Zero-centered gamma support in LayerNorm (LayerNorm1p)

### PR DESCRIPTION
**Pull request by ptrendx**
This PR adds a new option to LayerNorm (and other APIs that use LN) to initialize gamma parameter to 0 and use (1+gamma) instead of gamma.

This improves the precision characteristics of LN parameters and makes some techniques, like the weight decay, easier to implement for LN parameters.

### CODE REVIEW

**Comment by mikolajblaz** [comment]
@ptrendx I noticed that none of the backward kernels receive the `zero_centered_gamma` parameter, which is worrying. When computing the gradient wrt. `gamma` the `+1` should cancel out, but gradients wrt. all the other variables must receive a `gamma + 1` value.

From my limited understanding of kernels, those are the 2 places that might need a `+1`:
* https://github.com/NVIDIA/TransformerEngine/blob/742966ddc2ebd8ee49e5eef208e9ca86f17cffe3/transformer_engine/common/layer_norm/ln_bwd_kernels.cuh#L105
* https://github.com/NVIDIA/TransformerEngine/blob/742966ddc2ebd8ee49e5eef208e9ca86f17cffe3/transformer_engine/common/layer_norm/ln_bwd_kernels.cuh#L416

File: qa/L0_lint/pylintrc [hunk 1]
```diff
@@ -23,7 +23,8 @@ disable=too-many-locals,
         global-statement,
         too-many-branches,
         global-variable-not-assigned,
-        redefined-argument-from-local
+        redefined-argument-from-local,
+        too-many-function-args
```
Comment (ksivaman) [inline qa/L0_lint/pylintrc:27]: Why is this needed? Tests seem to pass w/o this

Comment (ptrendx) [reply qa/L0_lint/pylintrc:27]: This is for some of the autograd functions. For some you are working around this by passing `*args` lists, but not all have that treatment (and frankly it is not really any better than just passing the arguments directly).

Comment (ksivaman) [reply qa/L0_lint/pylintrc:27]: The `*args` is to handle different arguments for the onnx case, not to circumvent the linting problem.

Comment (ksivaman) [reply qa/L0_lint/pylintrc:27]: Okay, so if you look at the linter logs for when it threw an error for `too-many-function-args`, it was before your [Fix FP8 paths](https://github.com/NVIDIA/TransformerEngine/pull/67/commits/c36f3fbd2e031b212e545e93f39be0fb59d9f0c2) commit. So it was actually a useful message because the function was being incorrectly called with too many arguments.

Comment (ksivaman) [reply qa/L0_lint/pylintrc:27]: So we should remove from the `disable` list because it catches actual bugs.

Comment (ptrendx) [reply qa/L0_lint/pylintrc]: My bad, I thought about a different pylint warning. Reenabling.

File: transformer_engine/common/layer_norm/ln_api.cpp [hunk 2]
```diff
@@ -418,5 +423,66 @@ void nvte_layernorm_bwd(const NVTETensor dz,       // BxSxhidden_size
...
+                          NVTETensor dx,
+                          NVTETensor dgamma,
+                          NVTETensor dbeta,
+                          NVTETensor dgamma_part,
+                          NVTETensor dbeta_part,
+                          cudaStream_t stream,
+                          const int multiprocessorCount,
+                          NVTETensor workspace,
+                          NVTETensor barrier) {
+  NVTE_API_CALL(nvte_layernorm_bwd);
```
Comment (ksivaman) [inline transformer_engine/common/layer_norm/ln_api.cpp:471]:
```suggestion
  NVTE_API_CALL(nvte_layernorm1p_bwd);
```

File: transformer_engine/common/layer_norm/ln_bwd_kernels.cuh [hunk 3]
```diff
@@ -103,6 +103,9 @@ void ln_bwd_tuned_kernel(layer_norm::BwdParams params) {
                 compute_t x_tmp = x[it].data.elt[jt];
                 compute_t y_tmp = rs_r * (x_tmp - mu_r);
                 compute_t dy_tmp = compute_t(gamma[it].data.elt[jt]);
+                if (params.zero_centered_gamma) {
+                  dy_tmp += static_cast<compute_t>(1);
+                }
```
Comment (ksivaman) [inline transformer_engine/common/layer_norm/ln_bwd_kernels.cuh:108]: These lines (even for `zero_centered_gamma=False`) are causing the numerical changes.

File: transformer_engine/common/layer_norm/ln_bwd_kernels.cuh [hunk 4]
```diff
@@ -414,6 +417,9 @@ void ln_bwd_general_kernel(layer_norm::BwdParams params) {
                 compute_t x_ij = x.data.elt[jt];
                 compute_t y_ij = rs * (x_ij - mu);
                 compute_t g_ij = gamma[it].data.elt[jt];
+                if (params.zero_centered_gamma) {
+                  g_ij += static_cast<compute_t>(1);
+                }
```
Comment (ksivaman) [inline transformer_engine/common/layer_norm/ln_bwd_kernels.cuh:422]: Same as above comment

**Review by ksivaman — APPROVED** [review]
LGTM

**Comment by ksivaman** [comment]
ci errors are unrelated (ONNX)
